src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,107 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacityInternal(size + 1);
+        elementData[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index: " + index);
+        }
+        ensureCapacityInternal(size + 1);
+        System.arraycopy(elementData,index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        ensureCapacityInternal(size + list.size());
+        for (int i = 0; i < list.size(); i++) {
+            elementData[size++] = list.get(i);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeCheck(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeCheck(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheck(index);
+        T removedElement = (T) elementData[index];
+        int numMoved = size - index - 1;
+        if (numMoved > 0) {
+            System.arraycopy(elementData, index + 1, elementData, index, numMoved);
+        }
```

Comment:
arraycopy handles this fine
```suggestion
        
            System.arraycopy(elementData, index + 1, elementData, index, numMoved);
        
```